native/core/src/execution/datafusion/shuffle_writer.rs

```diff
@@ -1504,25 +1586,65 @@ mod test {
     #[test]
     #[cfg_attr(miri, ignore)] // miri can't call foreign function `ZSTD_createCCtx`
     fn test_insert_larger_batch() {
+        shuffle_write_test(10000, 1, 16, None);
+    }
+
+    #[test]
+    #[cfg_attr(miri, ignore)] // miri can't call foreign function `ZSTD_createCCtx`
+    fn test_insert_smaller_batch() {
+        shuffle_write_test(1000, 1, 16, None);
+        shuffle_write_test(1000, 10, 16, None);
+    }
+
+    #[test]
+    #[cfg_attr(miri, ignore)] // miri can't call foreign function `ZSTD_createCCtx`
+    #[cfg(not(target_os = "macos"))] // Github MacOS runner fails with "Too many open files".
```

Comment:
These tests fail on MacOS runners with "Too many open files" error. `ulimit` cannot help too.

I skip them on MacOS runners. We have ubuntu runners to test them.

Reply:
The test `shuffle_write_test(10000, 10, 200, Some(10 * 1024 * 1024))` spilled 1700 times, it spills too frequently for data of this size. Seems that the excessive spilling problem is inevitable if we reserve full batch capacity for the arrow builder.

Reply:
This PR seems like an important improvement because it now uses the memory pool features. Perhaps we can follow up with optimizations to reduce spilling. wdyt @Kontinuation?

Reply:
Sure. Let's merge this.

I'm also considering adding a native sort-based shuffle writer that works better with constraint resources.

Reply:
We've discussed to support sort-based shuffle in the native shuffle writer, similar to Spark shuffle in the early development. So I think it is on our roadmap though it is not urgent at that moment.